### PR TITLE
refactor: no longer write a config file by default

### PR DESCRIPTION
**Pull request by frrist**
### Why
This PR addresses #3004 by preventing a config file from being written to the bacalhau repo when the node is initialized. With this change, bacalhau will no longer write a config file in the bacalhau repo. 

The initial motivation for this work spawned from https://github.com/bacalhau-project/bacalhau/issues/3003 and https://github.com/bacalhau-project/bacalhau/pull/2999. When the demo networks were re-deployed the PeerIDs and IP addresses of the bacalhau bootstrap nodes and IPFS nodes changed. While the code's default values for these entities were updated accordingly, users with pre-existing Bacalhau nodes encountered issues. Their local config files continued to reference outdated bootstrap and swarm peer values, failing to align with the newly established defaults. (Reminder: values in a config file will override the default values as outlined in our [docs](https://docs.bacalhau.org/running-node/configuration#configuring-a-bacalhau-node)). Users with these outdated config values are unable to bootstrap to the network and/or retrieve content over the network.

This change will not fix issues previous users are experiencing as they still hold an invalid config. But it will prevent issues like this happening again in the since when we change defaults now users without a config will receive those new values when they update their version.

### Decision
**The decision we need to make here is: Do we want to follow this change with a PR that implements config file migration code?**

If we implement a migration:
- Users won't need to manually edit their config file if it contains outdated values, bacalhau will do it for them in this specific case.

If we don't implement a migration:
- Users with a config file will need to edit the file:
   - removing values of `Node.IPFS.SwarmPeers`.
   - removing values of `Node.BootstrapAddresses` if and only if they intend to use the `serve --peers=config` flag which forces bacalhau to use the bootstrap addresses defined in the config.

Migration v No Migration:
The migration for this would entail adding code to the bacalhau config package that explicitly checks the loaded config file for Bootstrap addresses and Swarm addresses such as these:
- https://github.com/bacalhau-project/bacalhau/blob/1a225c03c10cdb8fed7bd99b802b8e9d7a54ac8e/pkg/config/configenv/production.go#L59-L70
- https://github.com/bacalhau-project/bacalhau/blob/1a225c03c10cdb8fed7bd99b802b8e9d7a54ac8e/pkg/config/configenv/production.go#L33-L36
In the event they are present it will replace them with the correct values, otherwise it's a noop.

### CODE REVIEW

**Comment by dtbuchholz** [comment]
note: i can confirm this fixed the issue i opened with #3003. i manually removed `Node.IPFS.SwarmPeers` entirely and replaced the values for `Node.BootstrapAddresses` with those outlined above. tested with both `bacalhau` and `bacalhau devstack`—thanks for the quick resolution!

File: pkg/config/getters.go [hunk 1]
```diff
@@ -161,3 +164,144 @@ func GetStoragePath() string {
...
+	if err != nil {
+		return err
+	}
+
+	if isNested {
+		val, ok := compositeValue[key]
+		if !ok {
+			// NB(forrest): this case should never happen as we ensure all configuration values
+			// have a corresponding key via code gen. If this does occur it represents an error we need to debug.
+			return fmt.Errorf("CRITICAL ERROR: invalid configuration detected for key: %s. Config value not found", key)
```
Comment (rossjones) [inline pkg/config/getters.go:234]: Where this eventually unrolls to ForKey() the error we set here is ignored and "value not of expected type, got: %T" return instead.  Should we log the error as well if it is important?

File: pkg/config/getters.go [hunk 2]
```diff
@@ -161,3 +164,144 @@ func GetStoragePath() string {
...
+func Get[T any](key string) (T, error) {
+	raw := viper.Get(key)
+	if raw == nil {
+		return zeroValue[T](), fmt.Errorf("value not found for %s", key)
+	}
+
+	var val T
+	val, ok := raw.(T)
+	if !ok {
+		err := ForKey(key, &val)
```
Comment (rossjones) [inline pkg/config/getters.go:181]: There's a lot of code associated with this path through `Get[T]` do you have an ide of how often we expect to take this path?

**Review by rossjones — COMMENTED** [review]
I guess my only concern now is that we don't know what a config file looks like if there isn't a default one (or even a commented out one) to crib from.  Should we have something in https://docs.bacalhau.org/?  I went to take a look but couldn't find anything.   

> Do we want to follow this change with a PR that implements config file migration code?

I think yes, as long as it only replaces previously good (now bad) defaults.  I wonder whether we should have a general migration strategy so that we have a known way of providing migrations between releases?  e.g. specific code that is run going from 1.1.4 to 1.1.5, although I suspect this would be too big a burden unless we said it was only for 1.x.x (e.g. we have to maintain compat and remove nothing until we get to 2.x)

File: pkg/config/getters.go (hunk 2 above)
Comment (rossjones) [reply pkg/config/getters.go:181]: To answer my own question.  Of the 7 times this func (`Get`) is called, 6 times it is `Get[string]` and once it is `Get[bool]`.

**Comment by wdbaruni** [comment]
> I think yes, as long as it only replaces previously good (now bad) defaults. I wonder whether we should have a general migration strategy so that we have a known way of providing migrations between releases

That is my recommendation as well. We have different versioning for our repos not related to releases, and can be found under `~/.bacalhau/repo.version`. The current version is 1, and this change can bump the version to 2 or something similar. Having a general migration strategy between repo versions is my recommendation as well during my discussion with Forrest yesterday.

For this specific migration, we should load the existing config, compare each value with current and previous defaults, and only write back the config if the values don't match. The goal is not only to replace `Node.IPFS.SwarmPeers`, but delete all default values

**Comment by frrist** [comment]
> I guess my only concern now is that we don't know what a config file looks like if there isn't a default one (or even a commented out one) to crib from

Yeah, a concern I share. We can either continue to generate one as we do now, but with everything commented out (sshd, postgres, etc. do this). Or we can add a `bacalhau config` command that spits our the default config to stdout (e.g. `bacalhau config create > config.yaml`) . There is some appetite for a config API in https://github.com/bacalhau-project/bacalhau/issues/2941 so I am leaning towards the latter option. I am hesitant to put a default config in the docs at this point in time given how frequently we add new config values. Docs could quickly become outdated if we don't stay on top of config changes. Generating a default config on demand seems simpler to me.

File: pkg/config/getters.go (hunk 1 above)
Comment (frrist) [reply pkg/config/getters.go:234]: Great catch. I've adjusted the error message there to include the original error. I will also add a log line here.

**Comment by rossjones** [comment]
> we can add a bacalhau config command that spits our the default config to stdout (e.g. bacalhau config create

I think `bacalhau config defaults` might be useful, but I'm not sure about priority as I think it might be lower priority than other issues (such as the migration).  Is it worth putting into a ticket to get picked up later?

**Comment by wdbaruni** [comment]
> I think `bacalhau config defaults` might be useful, but I'm not sure about priority as I think it might be lower priority than other issues (such as the migration). Is it worth putting into a ticket to get picked up later?

I agree with Ross. Not a priority and better to wait until we get more signals on how useful it is

**Comment by frrist** [comment]
Anything blocking this from being approved? The code that adds a migration can follow this work.

**Review by rossjones — APPROVED** [review]

**Pull request merged**